src/cloud.c

```diff
@@ -73,76 +81,89 @@ static int map_routing_key_to_msg_type(const char *routing_key)
 {
-	struct cloud_msg msg;
-	json_object *jso;
-	bool consumed = true;
+	struct cloud_msg *msg = l_new(struct cloud_msg, 1);
 
-	jso = json_tokener_parse(body);
-	if (!jso) {
-		hal_log_error("Error on parse JSON object");
-		return consumed;
-	}
+	msg->type = map_routing_key_to_msg_type(routing_key);
 
-	msg.type = map_routing_key_to_msg_type(routing_key);
-
-	switch (msg.type) {
+	switch (msg->type) {
 	case UPDATE_MSG:
-		msg.device_id = parser_get_key_str_from_json_obj(jso, "id");
-		msg.list = parser_update_to_list(jso);
-		if (!msg.device_id || !msg.list) {
+		msg->device_id = parser_get_key_str_from_json_obj(jso, "id");
+		msg->list = parser_update_to_list(jso);
+		if (!msg->device_id || !msg->list) {
 			hal_log_error("Malformed JSON message");
-			goto done;
+			goto err;
 		}
 
 		break;
 	case REQUEST_MSG:
-		msg.device_id = parser_get_key_str_from_json_obj(jso, "id");
-		msg.list = parser_request_to_list(jso);
-		if (!msg.device_id || !msg.list) {
+		msg->device_id = parser_get_key_str_from_json_obj(jso, "id");
+		msg->list = parser_request_to_list(jso);
+		if (!msg->device_id || !msg->list) {
 			hal_log_error("Malformed JSON message");
-			goto done;
+			goto err;
 		}
 
 		break;
 	case REGISTER_MSG:
-		msg.device_id = parser_get_key_str_from_json_obj(jso, "id");
-		msg.token = parser_get_key_str_from_json_obj(jso, "token");
-		if (!msg.device_id || !msg.token) {
+		msg->device_id = parser_get_key_str_from_json_obj(jso, "id");
+		msg->token = parser_get_key_str_from_json_obj(jso, "token");
+		if (!msg->device_id || !msg->token) {
 			hal_log_error("Malformed JSON message");
-			goto done;
+			goto err;
 		}
 
 		break;
 	case UNREGISTER_MSG:
-		msg.device_id = parser_get_key_str_from_json_obj(jso, "id");
-		if (!msg.device_id) {
+		msg->device_id = parser_get_key_str_from_json_obj(jso, "id");
+		if (!msg->device_id) {
 			hal_log_error("Malformed JSON message");
-			goto done;
+			goto err;
 		}
 
 		break;
 	default:
 		hal_log_error("Unknown event %s", routing_key);
-		goto done;
+		goto err;
 	}
 
-	consumed = cloud_cb(&msg, user_data);
-done:
-	if (msg.type == UPDATE_MSG || msg.type == REQUEST_MSG)
-		l_queue_destroy(msg.list, l_free);
+	return msg;
+err:
+	cloud_msg_destroy(msg);
+	return NULL;
+}
 
+/**
+ * Callback function to consume and parse the received message from AMQP queue
+ * and call the respective handling callback function. In case of a error on
+ * parse, the message is consumed, but not used.
+ *
+ * Returns true if the message envelope was consumed or returns false otherwise.
+ */
+static bool on_cloud_receive_message(const char *exchange,
+				     const char *routing_key,
+				     const char *body, void *user_data)
+{
+	struct cloud_msg *msg;
+	bool consumed = true;
+	json_object *jso;
+
+	jso = json_tokener_parse(body);
+	if (!jso) {
+		hal_log_error("Error on parse JSON object");
+		return NULL;
+	}
+
+	msg = create_msg(routing_key, jso);
+	if (!msg)
+		goto free_jso;
```

Comment:
Is this goto necessary? It makes more sense to me to check if the msg is not NULL and execute the following code.
`consumed = cloud_cb(msg, user_data);
cloud_msg_destroy(msg);`

What do you think?

Reply:
Done